File: camerax_android/android/src/main/kotlin/dev/hebei/camerax_android/core/CameraControl.kt
package dev.hebei.camerax_android.core

import androidx.annotation.FloatRange
import kotlinx.coroutines.guava.await

class CameraControl internal constructor(val obj: androidx.camera.core.CameraControl) {

    suspend fun enableTorch(torch: Boolean) {
        obj.enableTorch(torch).await()
    }

    suspend fun setZoomRatio(ratio: Float) {
        obj.setZoomRatio(ratio).await()
    }

    suspend fun setLinearZoom(@FloatRange(from = 0.0, to = 1.0) linearZoom: Float) {
        obj.setLinearZoom(linearZoom).await()
    }

    suspend fun startFocusAndMetering(action: FocusMeteringAction): Boolean {
        val result = obj.startFocusAndMetering(action.obj).await()
        return result.isFocusSuccessful
    }

    suspend fun cancelFocusAndMetering() {
        obj.cancelFocusAndMetering().await()
    }

    suspend fun setExposureCompensationIndex(value: Int): Int {
        return obj.setExposureCompensationIndex(value).await()
    }
}

File: camerax_android/lib/src/legacy/core/camera_control.dart
import 'package:camerax_android/src/legacy/camerax.g.dart' as $native;
import 'package:camerax_platform_interface/camerax_platform_interface.dart'
    as $interface;

import 'focus_metering_action.dart';

final class CameraControl extends $interface.CameraControl {
  final $native.CameraControl obj;

  CameraControl.$native(this.obj) : super.impl();

  @override
  Future<void> cancelFocusAndMetering() async {
    await obj.cancelFocusAndMetering();
  }

  @override
  Future<void> enableTorch(bool torch) async {
    await obj.enableTorch(torch);
  }

  @override
  Future<int> setExposureCompensationIndex(int value) async {
    final newValue = await obj.setExposureCompensationIndex(value);
    return newValue;
  }

  @override
  Future<void> setLinearZoom(double linearZoom) async {
    await obj.setLinearZoom(linearZoom);
  }

  @override
  Future<void> setZoomRatio(double ratio) async {
    await obj.setZoomRatio(ratio);
  }

  @override
  Future<bool> startFocusAndMetering(
      $interface.FocusMeteringAction action) async {
    if (action is! FocusMeteringAction) {
      throw TypeError();
    }
    final result = await obj.startFocusAndMetering(action.obj);
    return result.isFocusSuccessful;
  }
}
